pyembed/src/importer.rs

```diff
@@ -900,7 +904,219 @@ impl OxidizedFinder {
+        // Compute _PathEntryFinder::package
+        let pkg = {
+            let current_exe = self.exe_realpath(py)?;
+            let not_exe_err = || {
+                let msg = format!(
+                    "path {} does not begin in \'{}\' (if you're sure it does, check Python's file-system encoding)", &path, current_exe.display());
+                let mut err = PyErr::new::<ImportError, _>(py, msg);
+                let exc = err.instance(py);
+                if let Err(setattr_err) = exc.setattr(py, "path", path.clone_ref(py)) {
+                    setattr_err
+                } else {
+                    err
+                }
+            };
+            let pbuf = pyobject_to_pathbuf(py, path.clone_ref(py))?;
+            let mut p: &std::path::Path = pbuf.as_ref();
+            // path could point "inside" current_exe. If we can't canonicalize,
+            // strip the last component and check again. (zipimporter does
+            // somethings similar)
+            let mut abs_p = p.canonicalize();
+            // Count the number of components at the end of `path` will be in pkg
+            let mut parts: usize = 0;
+            while abs_p.is_err() {
+                p = match p.parent() {
+                    Some(parent) => {
+                        parts += 1;
+                        parent
+                    }
+                    None => {
+                        return Err(not_exe_err());
+                    }
+                };
+                abs_p = p.canonicalize();
+            }
+            let abs_p = abs_p.unwrap();
+            if abs_p != current_exe {
+                return Err(not_exe_err());
+            }
+            // Extract the part of `path` after current_exe
+            // FIXME: This takes two allocations, but it shouldn't even use one.
+            let tail = pbuf
+                .iter()
+                .rev()
+                .take(parts)
+                .collect::<std::path::PathBuf>()
+                .iter()
+                .rev()
+                .collect::<std::path::PathBuf>();
+            _PathEntryFinder::parse_path_to_pkg(py, &tail)?
+        };
+        let finder = OxidizedFinder::create_instance(py, Arc::clone(self.state(py)))?;
+        // sys.path_hooks is only read after importlib._bootstrap_internal has
+        // been setup, either in that module or after initialization by
+        // PyImport_GetImporter checking for the existence of config->run_filename
+        // has been setup, so it's safe to import os now.
+        let paths = PyList::new(py, &[py.import("os")?.call(py, "fspath", (path,), None)?]);
+        _PathEntryFinder::create_instance(py, finder.into_object(), paths, pkg)
+    }
+}
+
+// A (mostly compliant) `importlib.abc.PathEntryFinder` that delegates paths
+// within the current executable to the `OxidizedFinder` whose `path_hook`
+// method created it. This is a private implementation detail.
+py_class!(class _PathEntryFinder |py| {
```

Comment:
Why is this prefixed with `_`? Perhaps we could rename it `OxidizedPathEntryFinder` so the name is consistent with other types defined by this Python extension?

Reply:
I'll change the name if you want, but allow me to explain my choices first.

I prefixed the name with `_` to mark `oxidized_importer._PathEntryFinder` as [private](https://www.python.org/dev/peps/pep-0008/#descriptive-naming-styles) from Python (none of the `py_class!`es in the module are [`pub`lic](http://dgrunwald.github.io/rust-cpython/doc/cpython/macro.py_class.html#syntax) from Rust). It's always possible to make a private API public later, but it's hard to change or remove a public API. My habit is to start off a feature with the smallest possible public API surface area. As my users and I gain experience with the feature, I add more public interfaces.

I eschewed prefixing the name with `Oxidized` for two reasons: (1) `_PathEntryFinder` is currently private, so Python code (other than pyembed's own tests) never needs to touch it; and (2) I generally prefer to [avoid stutter](https://blog.golang.org/package-names#TOC_3.) when naming interfaces.

Of course, like I said, I'll change the name however you want.

Reply:
If the type gets exposed to Python code, IMO it should be exported from the module to facilitate `isinstance()` checks and other type sniffing.

Reply:
Done. 8625a80d283d5c698d677854058bb1eec4b6509d